## Métricas de Evaluación

### Métricas de Regresión 

#### MSE: Error cuadrático medio - Mean Square Error

En estadística es un estimador que mide el cálculo del promedio de la diferencia de los errores cuadrados predichos o estimados y esperados para las etiquetas de un conjunto de datos. Matemáticamente se calcula de la siguiente forma:

$MSE = \frac{1}{n}\sum(\overset{\text{^}}{x}_i-x_i)^2$

Donde $n$ es el número de datos, $x_i$ representa los valores observados y $\overset{\text{^}}{x}$ los valores predichos.

<img src="MSE-Graph.PNG" width="350"/>
<img src="MSE-Graph2.PNG" width="350"/>

El MSE es una función que mide la probabilidad de que a una variable le ocurra cierta transformación ante ciertas condiciones esto corresponde al valor esperado de la perdida cuadrática, la diferencia se debe a la aleatoriedad de la muestra o porque el estimador carece de datos que le permita producir una estimación con mayor presición. 

In [17]:
#Cálculo de MSE con Sciki learn:

from sklearn.metrics import mean_squared_error

y_estimada = [3, -0.5, 2, 7]
y_predicha = [2.5, 0.0, 2, 8]
mean_squared_error(y_true, y_pred)

y_estimada = [3, -0.5, 2, 7]
y_predicha = [2.5, 0.0, 2, 8]
mean_squared_error(y_true, y_pred, squared=False)

y_estimada = [[0.5, 1],[-1, 1],[7, -6]]
y_predicha = [[0, 2],[-1, 2],[8, -5]]
mean_squared_error(y_true, y_pred)

array([0.41666667, 1.        ])

In [18]:
mean_squared_error(y_estimada, y_predicha, squared=False)

0.8227486121839513

In [19]:
mean_squared_error(y_estimada, y_predicha, multioutput='raw_values')

array([0.41666667, 1.        ])

In [20]:
mean_squared_error(y_estimada, y_predicha, multioutput=[0.3, 0.7])

0.825

In [23]:
#Cálculo con Pytorch

import torch
import torch.nn as nn

# define the input and target tensors
y_estimada = torch.tensor([[0.5, 1],[-1, 1],[7, -6]])
y_predicha = torch.tensor([[0, 2],[-1, 2],[8, -5]])

# print input and target tensors
print("y_estimada:\n", y_estimada)
print("y_predicha:\n", y_predicha)

# create a criterion to measure the mean squared error
mse = nn.MSELoss()

# compute the loss (mean squared error)
output = mse(y_estimada, y_predicha)

# output.backward()
print("MSE loss:", output)

y_estimada:
 tensor([[ 0.5000,  1.0000],
        [-1.0000,  1.0000],
        [ 7.0000, -6.0000]])
y_predicha:
 tensor([[ 0,  2],
        [-1,  2],
        [ 8, -5]])
MSE loss: tensor(0.7083)


https://www.tutorialspoint.com/how-to-measure-the-mean-squared-error-squared-l2-norm-in-pytorch

####  RMSE: Raíz de la Media Cuadrática - Root Mean Squared Error

Es normalmente usada para medir la diferencia entre los valores predichos por un modelo y los valores observados. Cuando el RMSE es calculado las unidades son iguales al valor target predicho, o tienen una misma escala. Matemáticamente se representa de la siguiente manera:

$RMSD = \sqrt\frac{\sum^N_{i=1}(x_i-\overset{\text{^}}{x})^2}{N}$

Donde $x_i$ es el valor esperado del conjunto de datos, $\overset{\text{^}}{x}$ es el valor predicho, expresado en unidades y no unidades cuadradadas a diferencia del MSE, entre $N$ que es el total de las muestras.

$RMSE = \sqrt{MSE} $

El RSMD sirve para mostrar un agregado de las magnitudes de los errores en las predicciones de los elementos identificables en un conjunto de datos a una medida de capacidad predictiva, por lo cual es una medida de exactitud para comparar errores de pronósticos  de diferentes modelos para un conjunto de datos particular y no entre conjuntos de datos. 

Ejemplo:

<img src="RMSE.PNG" width="450"/>

Las barras de error indican un intervalo de confianza del 95 % calculado con el error cuadrático medio del término de interacción de cada análisis de varianza. (REFERENCIA)

Un resultado perfecto de RMSE es 0.0 que significa que todas las predicciones son iguales a las esperadas o igual a un intervalo de confianza de 100%, lo cual no es usual que suceda e implica trivialidad en el modelo.

In [27]:
#Cálculo de RMSE con Sciki learn:

from sklearn.metrics import mean_squared_error
import math 

y_estimada = [3, -0.5, 2, 7]
y_predicha = [2.5, 0.0, 2, 8]
mean_squared_error(y_true, y_pred)

y_estimada = [3, -0.5, 2, 7]
y_predicha = [2.5, 0.0, 2, 8]
mean_squared_error(y_true, y_pred, squared=False)

y_estimada = [[0.5, 1],[-1, 1],[7, -6]]
y_predicha = [[0, 2],[-1, 2],[8, -5]]
mean_squared_error(y_true, y_pred)

0.7083333333333334

In [31]:
MSE = mean_squared_error(y_estimada, y_predicha, squared=False)
RMSE = math.sqrt(MSE)

print(RMSE)

0.9070549113388623


In [34]:
rv_MSE = mean_squared_error(y_estimada, y_predicha, multioutput=[0.3, 0.7])
rv_RMSE = math.sqrt(rv_MSE)

print(rv_RMSE)

0.9082951062292475


In [37]:
#Cálculo con Pytorch

import torch
import torch.nn as nn

# define the input and target tensors
y_estimada = torch.tensor([[0.5, 1],[-1, 1],[7, -6]])
y_predicha = torch.tensor([[0, 2],[-1, 2],[8, -5]])

# print input and target tensors
print("y_estimada:\n", y_estimada)
print("y_predicha:\n", y_predicha)

# create a criterion to measure the mean squared error
mse = nn.MSELoss()

# compute the loss (mean squared error)
output = mse(y_estimada, y_predicha)

RMSE = math.sqrt(output)

# output.backward()
print("RMSE loss:", RMSE)

y_estimada:
 tensor([[ 0.5000,  1.0000],
        [-1.0000,  1.0000],
        [ 7.0000, -6.0000]])
y_predicha:
 tensor([[ 0,  2],
        [-1,  2],
        [ 8, -5]])
RMSE loss: 0.8416253997266946


#### MAE: Error Medio Absoluto - Mean Absolute Error

El error medio absoluto es una técnica de medida que calcula la suma absoluta de errores dividos entre el tamaño total de la muestra. ¿Qué se considera "error"? Estadísticamente hablando, el error se refiere a la desviación de un valor observado con respecto a la media de una muestra. 

Por lo tanto MAE es una medida de errores entre observaciones aparejadas que describen un mismo fenómeno. En un sentido de una comparación entre los valores observados y predichos.

Se describe con la siguiente fórmula: 

$MAE = \frac{\sum^n_{1=i}{|y_i - x_i|}}{n} = \frac{\sum^n_{1=i}{|e_i|}}{n}$

Donde $y$ es la predicción y $x$ el valor verdadero, $n$ el total de la muestra. 

<img src="MAE1.PNG" width="450"/>

El anterior es una caracterización del pronóstico de errores y evaluación compativa de ejemplo de un pronóstico del uso de energía renobable, donde se pueden ver las varianzas de los errores, esto al ser divido por el total de la muestra nos permite comparar el dato, dado que se busca un solo resultado único con los valores reales.

Junto con el MSE y el RMSE son las métricas más populares para evaluar un modelo de regresión, pero a diferencia de las anteariores que castigan errores grandes, la MAE crece de manera líneal. Entre más cerca al 1 más exacto es el pronóstico en comparación al valor real, pero al exactitud total debería analizarse el tamaño de los datos o el diseño del modelo.

In [45]:
from sklearn.metrics import median_absolute_error

y_real = [3, -0.5, 2, 7]
y_pronosticado = [2.5, 0.0, 2, 8]
median_absolute_error(y_real, y_pronosticado)

0.5

In [73]:
y_real = [[0.5, 1], [-1, 1], [7, -6]]
y_pronosticado = [[0, 2], [-1, 2], [8, -5]]
median_absolute_error(y_real, y_pronosticado)

0.75

In [74]:
median_absolute_error(y_real, y_pronosticado, multioutput='raw_values')

array([0.5, 1. ])

In [75]:
median_absolute_error(y_real, y_pronosticado, multioutput=[0.3, 0.7])


0.85

In [56]:
#L1LOSS: Implementación con Pytorch

import torch
import torch.nn as nn

# define the input and target tensors
input = torch.randn(3, 4)
target = torch.randn(3, 4)

print("Input Tensor:\n", input)
print("Target Tensor:\n", target)

mae = nn.L1Loss()

output = mae(input, target)

print("MAE loss:", output)

Input Tensor:
 tensor([[-0.3015,  0.4134, -0.9815, -0.2512],
        [ 0.4303, -0.1585,  0.2871,  0.4258],
        [ 0.4995, -0.2316,  0.4205,  1.4265]])
Target Tensor:
 tensor([[-1.8159,  0.9825, -1.3881, -0.1789],
        [-0.9428,  1.6959,  0.2280,  1.1362],
        [ 0.1317,  1.1636,  0.2543, -0.3708]])
MAE loss: tensor(0.8572)


#### MAPE: Error Porcentual Absoluto Medio 

Dada que los problemas de regresión buscan encontrar un modelo donde se pueda compara los valores verdaderos y un pronóstico, que se puede representar como: $Z=(X,Y)$, donde se pueda medir que tan cerca está Y de X. Su representación matemática se ve así:

$MAPE = \frac{100\%}{n}\sum^n_{t=1}|\frac{A_t - F_t}{A_t}|$

Donde $A_t$ es el valor observado verdadero y $F_t$ es el valor pronosticado, su diferencia es dividida por $A_t$, el valor absoluto se presenta como un radio, que es sumado para cada uno de los valores pronosticados y divido por $n$.

En el siguiente ejemplo podemos ver el MAPE de varias acciones para varios modelos de regresión: 

<img src="MAPE.PNG" width="450"/>

Donde cada línea expresa los diversos errores porcentuales para cada acción espeficada. 

In [76]:
from sklearn.metrics import mean_absolute_percentage_error
y_real = [3, -0.5, 2, 7]
y_pronosticado = [2.5, 0.0, 2, 8]
mean_absolute_percentage_error(y_real, y_pronosticado)

0.3273809523809524

In [77]:
y_real = [[0.5, 1], [-1, 1], [7, -6]]
y_pronosticado = [[0, 2], [-1, 2], [8, -5]]
mean_absolute_percentage_error(y_real, y_pronosticado)

0.5515873015873016

In [78]:
mean_absolute_percentage_error(y_real, y_pronosticado, multioutput=[0.3, 0.7])


0.6198412698412699

In [79]:
# Cuando y_real es cero el valor puede ser elevadamente alto por lo cual se divide por epsilon

y_true = [1., 0., 2.4, 7.]
y_pred = [1.2, 0.1, 2.4, 8.]
mean_absolute_percentage_error(y_true, y_pred)

112589990684262.48

#### (R²): R cuadrado

### Métricas de Regresión 

#### Matriz de confusión

La matriz de confusión es una tabla que describe el rendimiento de un modelo de clasificación en un conjunto de datos de prueba cuyos valores verdaderos se conocen. Esta matriz tiene una dimensión de 2x2 con los valores reales en un eje y los predichos en otro.

<img src="https://www.juanbarrios.com/wp-content/uploads/2019/07/MATRIZ-CONFUSION-400x358.png" alt="Matriz de Confusión" style="height: 300px; width:300px;"/>

Como se puede observar en la imagen anterior, la matriz de confusión presenta 4 opciones que son los siguientes:
- Verdadero positivo (TP):  el modelo predice correctamente la clase positiva (la predicción y la realidad son ambas positivas). 
- Verdadero negativo (TN): el modelo predice correctamente la clase negativa (tanto la predicción como la realidad son negativas). 
- Falso negativo (FP):el modelo da una predicción errónea de la clase negativa (predicción-positiva, real-negativa). 
- Falso positivo (FN):el modelo predice erróneamente la clase positiva (predicho-negativo, real-positivo)

Los siguientes son ejemplos de las opciones:
- Verdadero positivo (TP):Una persna tiene cáncer y la prueba la demustra
- Verdadero negativo (TN): Una persona no tiene cáncer pero la prueba demuestra que no
- Falso negativo (FP): Un persona tiene cáncer pero la prueba demuestra que no
- Falso positivo (FN): La persona no tiene cáncer pero la prueba demuestra que si

<img src="https://images3.programmerclick.com/705/47/47733fd252edf6b399762ebe5eb71d31.png" alt="Matriz de Confusión" style="height: 300px; width:300px;"/>

A partir de las opciones anteriores surgen métricas de la matriz de confusión:
- Precisión (Precision) y exactitud (Accuracy)
- Sensibilidad (Recall) y especificidad (Especificity)

Precisión y exactitud 

Según la página Health Big Data, la precisión se refiere a la dispersión del conjunto de valores obtenidos a partir de mediciones repetidas de una magnitud. Cuanto menor es la dispersión mayor la precisión. 
Se representa por la proporción de verdaderos positivos dividido entre todos los resultados positivos, es decir la suma de los verdaderos positivos y falsos positivos. La siguiente imagen es la fórmula:

<img src="https://miro.medium.com/max/414/1*tHf0NHWGMwZTNe7TkylyYA.png" alt="Matriz de Confusión" style="height: 150px; width:150px;"/>

Por otro lado, la exactitud está relacionada con el sesgo de una estimación y se refiere que tan cerca está el resultado de una medición del valor verdadero. Se representa como la proporción de resultados verdaderos (tanto verdaderos positivos (VP) como verdaderos negativos (VN)) dividido entre el número total de casos examinados. La siguiente imagen es la fórmula:

<img src="https://live.staticflickr.com/65535/48051257888_760554732e_b.jpg" alt="Matriz de Confusión" style="height: 300px; width:450px;"/>

###### Sensibilidad (Recall) Especificidad (Especificity)

La sensibilidad nos indica el porcentaje que son positivos predicjos del total de positivos, es como verdaderos positivos como se puede observar en la siguiente imagen: 

<img src="https://miro.medium.com/max/444/1*ynGz0zu8ZlpgZ5voYRKfLg.png" alt="Matriz de Confusión" style="height: 150px; width:150px;"/>

En el caso de la espeficidad se refiere los casos negativos que fueron calificados correctamente.

<img src="https://geekymedics.com/wp-content/uploads/2018/06/Specificity-equation.jpg" alt="Matriz de Confusión" style="height: 200px; width:200px;"/>

###### Ejemplo de matriz de confusión

El siguiente es un ejemplo simple que fue implementado con Sklearn, en este caso se usó dos arreglos, uno con datos actuales/reales y otro con datos predecidos alambrados.Al llamar la función confusion_matrix con los dos arreglos obtenemos la respuesta en formato de matriz:

In [65]:
from sklearn.metrics import confusion_matrix
Valoes_reales = ['F','T','T','F','T','F','T','T','F','T','F','T']
Valores_pred = ['F','F','T','F','T','F','F','T','F','T','T','F']
confusion_matrix(Valoes_reales,Valores_pred)

array([[4, 1],
       [3, 4]], dtype=int64)

Al realizar los cálculos manualmente se puede decir que la matriz mostrada anteriormente fue calculada correctamente, y consideramos que es una forma muy simple de comprobar como está el modelo para predecir los datos.

#### F1 Score

Según diferentes paǵinas del internet, F1 es una propuesta de mejora de dos métricas de rendimiento más sencillas.  El F1 combina las medidas más comunes de precision y recall en un sólo valor, esto hace que el proceso sea más simple de poder realizar comparaciones combinando las medidas. Esta métrica ha sido diseñad para trabajar con datos desequilibrados.

###### F1 Score Formula

La siguiente es la fórmula de F1 realizando combinación entre la precisión y recall. 

<img src="https://lawtomated.com/wp-content/uploads/2019/10/F1-Score.png" alt="Matriz de Confusión" />

Se debe considedrar los siguientes puntos al utilizar F1 Score para evaluar un modelo:
- Sí la precisión y recall son altas entonces el modelo obtendrá una puntuación F1 alta 
- Sí la precisión y recall son bajas entonces el modelo obtendrá una puntuación F1 baja 
- Sí la precisión es baja y recall alta entonces el modelo obtendrá una puntuación F1 media 

###### Precisión

Precisión puede ser una métrica individual para evaluar modelo pero es utilizada y parte de F1 score. La siguiente es la fórmula de precisón:

<img src="https://miro.medium.com/max/622/1*hIPkZ3Su21sUdwi_A-arUg.png" alt="Matriz de Confusión" />

Según Joos Korstanje, la fórmula de precisión se puede interpretar que todo lo que se ha predicho como positivo, la precisión cuenta el porcentaje que es correcto:
- Un modelo no preciso puede encontrar muchos de los positivos, pero su método de selección es ruidoso: también detecta erróneamente muchos positivos que no son realmente positivos.
- Un modelo preciso es muy "puro": puede que no encuentre todos los positivos, pero los que el modelo clasifica como positivos tienen muchas probabilidades de ser correctos

###### Recall

Igual que precisón, la sensibilidad que se conoce como recall puede ser una métrica individual pero es un componente fundamental para el cálculo de F1. La siguiente es la fórmula de recall:

<img src="https://miro.medium.com/max/612/1*8kzRR5Pl1SI2tRUa3CvNsA.png" alt="Matriz de Confusión" />

Un modelo que tiene un resultado de recall alto logra encontrar todos los casos positivos de los datos, pero también se puede suceder que algunos casos negativos que los identifican como positivos. Del mismo modo, si el modelo que tiene un resultado de recall bajo no es capaz de enfcontrar la mayoría de los casos positivos en los datos. 

Un ejemplo en la vida real, normalmente cuando compramos frutas hay ocasiones que hay partes que están podridas, y al vendor sólo le interesa asegurar de que los clientes devuelven las frutas podridas y no le importa si los clientes devuelven frutas no podridas, es decir que la precisión no es tan importante en este caso. 

###### Ejemplo de F1 Score

El siguiente es un ejemplo simple que fue implementado con Sklearn, en este caso se usó dos arreglos, uno con datos actuales/reales y otro con datos predecidos alambrados.Al llamar la función f1_score con los dos arreglos obtenemos 0.6666 como respuesta.


In [68]:
from sklearn.metrics import f1_score
Valoes_reales = ['F','T','T','F','T','F','T','T','F','T','F','T']
Valores_pred = ['F','F','T','F','T','F','F','T','F','T','T','F']
f1_score(Valoes_reales,Valores_pred,average='macro')

0.6666666666666666

Como se ha mencionado anteriormente, si el resultado de F1 score es alto es mejor, pero igual en este caso estamos usando datos alambrados de una sola dimensión con una cantidad pequeña de datos, pero el resultado obtenido debe depender del problema de predicción para poder decir con certeza de sí es bueno o malo.

#### Accuracy (Exactitud)

La accuracy que se conoce como exactitud es otra métrica de clasificación más conocida ya que esta puede ser utilizada individualmente o ser parte de otras métricas. Esta métrica es muy simple y puede ser utilizada tanta para los problemas de clasificación binarios como a los multiclase. Se recomienda utilizar la métrica exactitud para los problemas de clasificación que están equilibrados y no sesgados o no hay desequilibrio de clases.Como se puede ver en la siguiente fórmula, la exactitud es la proporción de resultados verdaderos entre el número total de casos.

<img src="https://live.staticflickr.com/65535/48051257888_760554732e_b.jpg" alt="Matriz de Confusión"/>

###### Ejemplo de Recall

El siguiente es un ejemplo simple que fue implementado con Sklearn, en este caso se usó dos arreglos, uno con datos actuales/reales y otro con datos predecidos alambrados.Al llamar la función accuracy_score con los dos arreglos obtenemos 0.6666 como respuesta.

In [69]:
from sklearn.metrics import accuracy_score
Valoes_reales = ['F','T','T','F','T','F','T','T','F','T','F','T']
Valores_pred = ['F','F','T','F','T','F','F','T','F','T','T','F']
accuracy_score(Valoes_reales,Valores_pred)

0.6666666666666666

Como ya se ha mencionada anteriormente, la accuracy es una métrica representa el número de casos de datos correctamente clasificados sobre el número total de casos de datos.En este caso obtuvimos que casi 67% de las predicciones fueron correctas. Esta métrica mientras más alto sea el valor es mejor. 

#### Recall (Exhaustividad)

El recall es una métrica que trata de capturar el mayor número posible de positivos. Un ejemplo que puede utilizar esta métrica es construir un modelo que puede predecir si la persona tiene Covid o no, ya que se requiere capturar aunque no estemos seguros ya que uno puede ser asintómatico. Es decir que Recall da una medida con la que el modelo es capaz de identificar los datos relevantes.

<img src="https://miro.medium.com/max/612/1*8kzRR5Pl1SI2tRUa3CvNsA.png" alt="Matriz de Confusión" />

###### Ejemplo de recall

El siguiente es un ejemplo simple que fue implementado con Sklearn, en este caso se usó dos arreglos, uno con datos actuales/reales y otro con datos predecidos alambrados.Al llamar la función recall_score con los dos arreglos obtenemos 0.6666 como respuesta.

In [71]:
from sklearn.metrics import recall_score
Valoes_reales = ['F','T','T','F','T','F','T','T','F','T','F','T']
Valores_pred = ['F','F','T','F','T','F','F','T','F','T','T','F']
recall_score(Valoes_reales,Valores_pred,average='macro')

0.6857142857142857

#### Precision (Precisión)

Según el internet la métrica precisión sirve para realizar clasificación de los datos desquilibrados.Como ya se ha mencionado anteriormente, precisión puede ser una métrica individual y también ser componente de otras métricas como F1 Score y Matriz de confusión. La precisión es una métrica que cuantifica el número de predicciones positivas correctas obtenidas. Se calcula como la proporción de ejemplos positivos predichos correctamente dividida por el número total de ejemplos positivos que se predijeron.

Se recomienda utilizar la métrica precisión cuando se requiere estar muy seguro de la predicción. Un ejemplo real, es que cuando hay un empleado nuevo en la empresa se debe asignarle los accesos correctos para no romper ninguna política de seguridad. Este métrica puede ser utilizada tanto para la clasificación binaria como clasificación multiclases.

###### Precisión para la clasificación binaria

La precisión se calcula como el número de verdaderos positivos dividido por el número total de verdaderos positivos y falsos positivos. El rango del resultado es un valor entre 0 para una precisión nula y 1 para una precisión total o perfecta, es decir mientras más alto sea el valor el resultado será mejor. La siguiente es la fórmula de precisón:

<img src="https://miro.medium.com/max/622/1*hIPkZ3Su21sUdwi_A-arUg.png" alt="Matriz de Confusión" />

###### Precisión para la clasificación multiclases

La precisión es una métrica muy flexible ya que nos permite evaluar modelo que son de clasificación multiclase, y esta  se calcula como la suma de verdaderos positivos en todas las clases dividida por la suma de verdaderos positivos y falsos positivos en todas las clases.

###### Ejemplo de precisión

El siguiente es un ejemplo simple que fue implementado con Sklearn, en este caso se usó dos arreglos, uno con datos actuales/reales y otro con datos predecidos alambrados.Al llamar la función precision_score con los dos arreglos obtenemos 0.6666 como respuesta.

In [72]:
from sklearn.metrics import precision_score
Valoes_reales = ['F','T','T','F','T','F','T','T','F','T','F','T']
Valores_pred = ['F','F','T','F','T','F','F','T','F','T','T','F']
precision_score(Valoes_reales,Valores_pred,average='macro')

0.6857142857142857

Como ya se ha mencionada anteriormente, la precisión es una métrica de evalua las predicciones positivas realizadas que son correctas.En este caso obtuvimos que casi 69% de las predicciones fueron correctas.

Referencias.

AWS. (s.f.). AWS > Documentation > Amazon Machine Learning. Obtenido de https://docs.aws.amazon.com/es_es/machine-learning/latest/dg/evaluating-model-accuracy.html

Brownlee, J. (s.f.). Machine Learning Mastery. Obtenido de Regression Metrics for Machine Learning.

Research Gate. (s.f.). Figure 5 - uploaded by Alinda Friedman. Obtenido de https://www.researchgate.net/figure/Root-mean-square-error-RMSE-from-the-center-of-the-circle-for-the-one-glyph-task-in_fig3_309026476

Research Gate. (s.f.). Research Gate. Obtenido de https://www.researchgate.net/figure/Comparison-graph-of-model-performance-in-the-Mean-Absolute-Percentage-Error-MAPE_fig2_343349809

Sanjuán, F. J. (s.f.). Economipedia. Obtenido de R cuadrado ajustado (Coeficiente de determinación ajustado): https://economipedia.com/definiciones/r-cuadrado-ajustado-coeficiente-de-determinacion-ajustado.html#:~:text=El%20R%20cuadrado%20ajustado%20(o,en%20explicar%20la%20variable%20dependiente.

Science Direct. (s.f.). Obtenido de Mean Absolute Error: https://www.sciencedirect.com/topics/engineering/mean-absolute-error

Scikit Learn. (s.f.). API Reference. Obtenido de https://scikit-learn.org/stable/modules/classes.html#regression-metrics

Sitio Big Data. (s.f.). Aprendizaje automático de métricas de regresión (MSE). Obtenido de https://sitiobigdata.com/2019/05/27/modelos-de-machine-learning-metricas-de-regresion-mse-parte-2/#

Statistics How To. (s.f.). Absolute Error & Mean Absolute Error (MAE). Obtenido de https://www.statisticshowto.com/absolute-error/

Statistics How To. (s.f.). Mean Absolute Percentage Error (MAPE). Obtenido de https://www.statisticshowto.com/mean-absolute-percentage-error-mape/

Statologos. (s.f.). Error porcentual absoluto medio (MAPE). Obtenido de https://statologos.com/mapa-de-error-de-porcentaje-absoluto-medio/

Tutorials Point. (s.f.). Tutorials Point. Obtenido de How to measure the mean squared error(squared L2 norm) in PyTorch?: https://www.tutorialspoint.com/how-to-measure-the-mean-squared-error-squared-l2-norm-in-pytorch

Health Big Data. (2019, 26 de julio). La matriz de confusión y sus métricas. https://www.juanbarrios.com/la-matriz-de-confusion-y-sus-metricas/

Python Guides. (2022, 1 de febrero). Scikit learn Confusion Matrix. https://pythonguides.com/scikit-learn-confusion-matrix/

Joos Korstanje. (2021, 31 de agosto). The F1 score. https://towardsdatascience.com/the-f1-score-bec2bbc38aa6

scikit learn. (s.f.). F1_Score. https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html

AskPython. (s.f.). Precision and Recall in Python. https://www.askpython.com/python/examples/precision-and-recall-in-pythonc

Jason Brownlee. (2020, 3 de enero). How to Calculate Precision, Recall, and F-Measure for Imbalanced Classification. https://machinelearningmastery.com/precision-recall-and-f-measure-for-imbalanced-classification/

Rahul Agarwal. (2019, 17 de septiembre). The 5 Classification Evaluation metrics every Data Scientist must know.https://towardsdatascience.com/the-5-classification-evaluation-metrics-you-must-know-aa97784ff226 